dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,261 @@
+//
+// ETHSwap cannot be used by other contracts or by a third party mediating
+// the swap or multisig wallets.
+//
+// This code should be verifiable as resulting in a certain on-chain contract
+// by compiling with the correct version of solidity and comparing the
+// resulting byte code to the data in the original transaction.
+contract ETHSwap {
+    bytes4 private constant TRANSFER_FROM_SELECTOR = bytes4(keccak256("transferFrom(address,address,uint256)"));
+    bytes4 private constant TRANSFER_SELECTOR = bytes4(keccak256("transfer(address,uint256)"));
+    // Step is a type that hold's a contract's current step. Empty is the
+    // uninitiated or null value.
+    enum Step { Empty, Filled, Redeemed, Refunded }
+
+    struct Status {
+        Step step;
+        bytes32 secret;
+        uint256 blockNumber;
+    }
+
+    bytes32 constant RefundRecord = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+    bytes32 constant RefundRecordHash = 0xAF9613760F72635FBDB44A5A0A63C39F12AF30F950A6EE5C971BE188E89C4051;
+
+    // swaps is a map of contract hashes to the "swap record". The swap record
+    // has the following interpretation.
+    //   if (record == bytes32(0x00)): contract is uninitiated
+    //   else if (uint256(record) < block.number && sha256(record) != contract.secretHash):
+    //      contract is initiated and redeemable by the participant with the secret.
+    //   else if (sha256(record) == contract.secretHash): contract has been redeemed
+    //   else if (record == RefundRecord): contract has been refunded
+    //   else: invalid record. Should be impossible by construction
+    mapping(bytes32 => bytes32) public swaps;
+
+    // Vector is the information necessary for initialization and redemption
+    // or refund. The Vector itself is not stored on-chain. Instead, a key
+    // unique to the Vector is generated from the Vector data and keys
+    // the swap record.
+    struct Vector {
+        bytes32 secretHash;
+        uint256 value;
+        address initiator;
+        uint64 refundTimestamp;
+        address participant;
+    }
+
+    // contractKey generates a key hash which commits to the contract data. The
+    // generated hash is used as a key in the swaps map.
+    function contractKey(address token, Vector calldata v) public pure returns (bytes32) {
+        return sha256(
+            bytes.concat(
+                v.secretHash,
+                bytes20(v.initiator),
+                bytes20(v.participant),
+                bytes32(v.value),
+                bytes8(v.refundTimestamp),
+                bytes20(token)
+            )
+        );
+    }
+
+    // Redemption is the information necessary to redeem a Vector. Since we
+    // don't store the Vector itself, it must be provided as part of the
+    // redemption.
+    struct Redemption {
+        Vector v;
+        bytes32 secret;
+    }
+
+    function secretValidates(bytes32 secret, bytes32 secretHash) public pure returns (bool) {
```

Comment:
nit: I would prefer `canSecreteBeValidated` or even `isSecretValidated` for readabilty